engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/rollingformula/BaseRollingFormulaOperator.java

```diff
@@ -0,0 +1,271 @@
+        inputVectorType = getVectorType(inputColumnType);
+
+        // Re-use the formula column if it's already been created for this type. No need to synchronize; these
+        // operators are created serially.
+        formulaColumn = formulaColumnMap.computeIfAbsent(inputColumnType, t -> {
+            final FormulaColumn tmp = FormulaColumn.createFormulaColumn(outputColumnName,
+                    FormulaUtil.replaceFormulaTokens(formula, paramToken, PARAM_COLUMN_NAME));
+
+            final ColumnDefinition<?> inputColumnDefinition = ColumnDefinition
+                    .fromGenericType(PARAM_COLUMN_NAME, inputVectorType, inputColumnType);
+            tmp.initDef(Collections.singletonMap(PARAM_COLUMN_NAME, inputColumnDefinition));
+            return tmp;
+        });
+    }
+
+    protected BaseRollingFormulaOperator(
+            @NotNull final MatchPair pair,
+            @NotNull final String[] affectingColumns,
+            @Nullable final String timestampColumnName,
+            final long reverseWindowScaleUnits,
+            final long forwardWindowScaleUnits,
+            final Class<?> inputVectorType,
+            @NotNull final Map<Class<?>, FormulaColumn> formulaColumnMap,
+            @NotNull final TableDefinition tableDef) {
+        super(pair, affectingColumns, timestampColumnName, reverseWindowScaleUnits, forwardWindowScaleUnits, true);
+        this.formulaColumnMap = formulaColumnMap;
+        this.tableDef = tableDef;
+
+        final Class<?> columnType = tableDef.getColumn(pair.rightColumn).getDataType();
+        this.inputVectorType = inputVectorType;
+
+        // Re-use the formula column already created for this type.
+        formulaColumn = formulaColumnMap.computeIfAbsent(columnType, t -> {
+            throw new IllegalStateException("formulaColumnMap should have been populated for " + columnType);
+        });
+    }
+
+    @Override
+    public void initializeSources(@NotNull final Table source, @Nullable final RowRedirection rowRedirection) {
+        this.rowRedirection = rowRedirection;
+
+        if (rowRedirection != null) {
+            // region create-dense
+            maybeInnerSource = ArrayBackedColumnSource.getMemoryColumnSource(0, formulaColumn.getReturnedType());
+            // endregion create-dense
+            outputSource = WritableRedirectedColumnSource.maybeRedirect(rowRedirection, maybeInnerSource, 0);
+        } else {
+            maybeInnerSource = null;
+            // region create-sparse
+            outputSource = SparseArrayColumnSource.getSparseMemoryColumnSource(0, formulaColumn.getReturnedType());
+            // endregion create-sparse
+        }
+
+        primitiveOutputSource = ReinterpretUtils.maybeConvertToWritablePrimitive(outputSource);
+
+        outputChunkType = primitiveOutputSource.getChunkType();
+    }
+
+    private static Class<?> getVectorType(final Class<?> type) {
```

Comment:
I think you want to use `io.deephaven.vector.VectorFactory.forElementType()`, which gives you a `vectorType()`.